packages/replicache/src/kv/sqlite-store.ts

```diff
@@ -0,0 +1,376 @@
+import {RWLock} from '@rocicorp/lock';
+import type {ReadonlyJSONValue} from '../../../shared/src/json.ts';
+import {
+  promiseUndefined,
+  promiseVoid,
+} from '../../../shared/src/resolved-promises.ts';
+import {deepFreeze} from '../frozen-json.ts';
+import type {Read, Store, Write} from './store.ts';
+
+/**
+ * A SQLite prepared statement.
+ *
+ * `run` executes the statement with optional parameters.
+ * `all` executes the statement and returns the result rows.
+ * `finalize` releases the statement.
+ */
+export interface PreparedStatement {
+  run(...params: unknown[]): void;
+  all<T>(...params: unknown[]): T[];
+  finalize(): void;
+}
+
+export interface SQLiteDatabase {
+  /**
+   * Close the database connection.
+   */
+  close(): void;
+
+  /**
+   * Destroy or delete the database (e.g. delete file).
+   */
+  destroy(): void;
+
+  /**
+   * Prepare a SQL string, returning a statement you can execute.
+   * E.g. `const stmt = db.prepare("SELECT * FROM todos WHERE id=?");`
+   */
+  prepare(sql: string): PreparedStatement;
+
+  /**
+   * Check if the database is in a transaction.
+   */
+  isInTransaction(): boolean;
+}
+
+type SQLitePreparedStatements = {
+  begin: PreparedStatement;
+  beginImmediate: PreparedStatement;
+  commit: PreparedStatement;
+  rollback: PreparedStatement;
+
+  savepoint: PreparedStatement;
+  release: PreparedStatement;
+
+  get: PreparedStatement;
+  put: PreparedStatement;
+  del: PreparedStatement;
+};
+
+const getPreparedStatementsForSQLiteDatabase = (
+  db: SQLiteDatabase,
+): SQLitePreparedStatements => ({
+  begin: db.prepare('BEGIN'),
+  beginImmediate: db.prepare('BEGIN IMMEDIATE'),
+  commit: db.prepare('COMMIT'),
+  rollback: db.prepare('ROLLBACK'),
+
+  // Similar to https://github.com/WiseLibs/better-sqlite3/blob/674ce6be68a26742d9e24f8672da7888cea0aebb/lib/methods/transaction.js#L35-L39
+  savepoint: db.prepare('SAVEPOINT `\t_rc.\t`'),
+  release: db.prepare('RELEASE `\t_rc.\t`'),
+
+  get: db.prepare('SELECT value FROM entry WHERE key = ?'),
+  put: db.prepare('INSERT OR REPLACE INTO entry (key, value) VALUES (?, ?)'),
+  del: db.prepare('DELETE FROM entry WHERE key = ?'),
+});
+
+const rwLocks = new Map<string, RWLock>();
```

Comment:
It's unfortunate to need multiple levels of locking but I guess it is needed so we dont' have to create a connection per read/write.

Reply:
I do not think we need a global map of locks though. Each SQLiteStore can own a connection to SQLite which uses SQLite's built-in file based locking. We don't expect lots of instances of SQLiteStore for the same name (or really any?) in the same js context. Each SQLiteStore can then also own a RWLock for in-memory queuing.

Reply:
Yeah that makes sense. This was for safety but I will remove if there's that guarantee. Yeah, exactly, the alternative was spinning up many connections.